# Kaggle: Predict Loan Payback — Exploratory Data Analysis (EDA)

**Project:** Kaggle - Predict Loan Payback (Playground Series S5E11)
**Notebook:** `01_eda.ipynb`
**Author:** Brice [Last Name Optional]
**Organization:** Kaggle Series | Brice Machine Learning Projects
**Date Created:** November 2, 2025
**Last Updated:** November 2, 2025

---

### **Purpose**
Perform Exploratory Data Analysis (EDA) to understand the dataset used in the Kaggle “Predict Loan Payback” competition.
This notebook focuses on identifying data quality issues, understanding feature distributions, and uncovering relationships that may guide feature engineering and model selection.

---

### **Contents**
1. Imports & Setup
2. Data Loading
3. Initial Exploration
4. Missing Value Analysis
5. Univariate & Bivariate Analysis
6. Correlation Heatmaps
7. Observations & Next Steps


## Imports & Setup

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns


## Data Loading

In [3]:
loan_train_df = pd.read_csv('../data/raw/train.csv')
loan_test_df = pd.read_csv('../data/raw/test.csv')


## Initial Exploration

In [4]:
# Shape
loan_train_df.shape

(593994, 13)

In [5]:
# Information
loan_train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 593994 entries, 0 to 593993
Data columns (total 13 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   id                    593994 non-null  int64  
 1   annual_income         593994 non-null  float64
 2   debt_to_income_ratio  593994 non-null  float64
 3   credit_score          593994 non-null  int64  
 4   loan_amount           593994 non-null  float64
 5   interest_rate         593994 non-null  float64
 6   gender                593994 non-null  object 
 7   marital_status        593994 non-null  object 
 8   education_level       593994 non-null  object 
 9   employment_status     593994 non-null  object 
 10  loan_purpose          593994 non-null  object 
 11  grade_subgrade        593994 non-null  object 
 12  loan_paid_back        593994 non-null  float64
dtypes: float64(5), int64(2), object(6)
memory usage: 58.9+ MB


In [6]:
# Head
loan_train_df.head()

id  annual_income  debt_to_income_ratio  credit_score  loan_amount  \
0   0       29367.99                 0.084           736      2528.42   
1   1       22108.02                 0.166           636      4593.10   
2   2       49566.20                 0.097           694     17005.15   
3   3       46858.25                 0.065           533      4682.48   
4   4       25496.70                 0.053           665     12184.43   

   interest_rate  gender marital_status education_level employment_status  \
0          13.67  Female         Single     High School     Self-employed   
1          12.92    Male        Married        Master's          Employed   
2           9.76    Male         Single     High School          Employed   
3          16.10  Female         Single     High School          Employed   
4          10.21    Male        Married     High School          Employed   

         loan_purpose grade_subgrade  loan_paid_back  
0               Other             C3             1.0  
1  Debt consolidation             D3             0.0  
2  Debt consolidation             C5             1.0  
3  Debt consolidation             F1             1.0  
4               Other             D1             1.0

### Determine Null Values in Dataset

In [7]:
loan_train_df.isnull().sum().sort_values(ascending=False)

id                      0
annual_income           0
debt_to_income_ratio    0
credit_score            0
loan_amount             0
interest_rate           0
gender                  0
marital_status          0
education_level         0
employment_status       0
loan_purpose            0
grade_subgrade          0
loan_paid_back          0
dtype: int64

### Diagnostic Check

- The dataset shows no nulls
- The dataset has 13 features
- The dataset has and almost 594K rows
- It seems unlikely there are no null values.
- It is possible there may be strings of `NA`, `N/A`, `None`, `Missing`, etc
- If the columns have such entries, it would not show up as a null value.
- The data needs to be examined more to determine the null values

In [8]:
# Quick Diagnostic Check
print("Dataset Shape:", loan_train_df.shape)
print("Object columns:", loan_train_df.select_dtypes(include='object').columns.tolist())

for col in loan_train_df.select_dtypes(include='object').columns:
    print(f"\nColumn: {col}")
    print(loan_train_df[col].value_counts(dropna=False).head(10))


Dataset Shape: (593994, 13)
Object columns: ['gender', 'marital_status', 'education_level', 'employment_status', 'loan_purpose', 'grade_subgrade']

Column: gender
gender
Female    306175
Male      284091
Other       3728
Name: count, dtype: int64

Column: marital_status
marital_status
Single      288843
Married     277239
Divorced     21312
Widowed       6600
Name: count, dtype: int64

Column: education_level
education_level
Bachelor's     279606
High School    183592
Master's        93097
Other           26677
PhD             11022
Name: count, dtype: int64

Column: employment_status
employment_status
Employed         450645
Unemployed        62485
Self-employed     52480
Retired           16453
Student           11931
Name: count, dtype: int64

Column: loan_purpose
loan_purpose
Debt consolidation    324695
Other                  63874
Car                    58108
Home                   44118
Education              36641
Business               35303
Medical                22806
Vacati

---

## 🧩 Missing Value Analysis — Summary

A thorough null-value diagnostic was conducted to assess data completeness and ensure data quality before encoding or feature engineering.

### **Steps Performed**
1. Checked for missing values using `.isnull().sum()` across all columns.
2. Verified the absence of "fake nulls" (e.g., `"NA"`, `"Unknown"`, empty strings) by reviewing unique categorical values.
3. Confirmed dataset shape `(593,994 rows × 13 columns)` and data types using `.info()`.

### **Findings**
- No missing values were detected across any columns (`.isnull().sum()` returned zero).
- No placeholder or “fake” missing entries (e.g., `"N/A"`, `"Unknown"`, `""`) were found in the categorical columns.
- Each categorical feature contained clean, interpretable categories such as gender, marital status, education level, and employment status.

### **Reasoning & Interpretation**
The Kaggle *Playground Series (S5E11)* dataset appears to be **synthetic and pre-cleaned**, which aligns with the competition’s intent to emphasize **modeling and feature exploration** rather than data cleaning.
Although it is uncommon for real-world financial datasets to contain zero missing values, this design ensures consistent starting conditions for all competitors.

### **Next Step**
Since data completeness has been confirmed, the next logical phase is to:
- Explore categorical feature distributions.
- Determine appropriate encoding strategies (binary, one-hot, or ordinal).
- Proceed toward feature correlation and model preparation.

---


---

## 🧱 Categorical Feature Exploration

With data completeness confirmed, the next step involves examining categorical feature distributions to understand the dataset’s structure and potential predictive value.

### **Objectives**
1. Identify category frequencies and detect class imbalance.
2. Understand the relative representation of borrower demographics and loan characteristics.
3. Determine which features are binary, nominal, or ordinal — guiding the encoding strategy for model readiness.

### **Approach**
- Generate normalized frequency tables for each categorical feature to visualize proportional distributions.
- Use bar plots to visualize key categorical variables such as `gender`, `marital_status`, `education_level`, and `loan_purpose`.
- Record insights about potential feature importance or bias (e.g., dominant borrower profiles, rare categories).

### **Rationale**
Understanding categorical balance is essential in financial risk modeling.
Uneven distributions can:
- Introduce bias into classification models.
- Distort model performance metrics if not addressed through resampling or regularization.
- Influence the choice between **label encoding**, **one-hot encoding**, or **ordinal encoding**.

---


---

## 🧩 Summary & Key Insights

**Overview:**
Summarize the key findings from the EDA. Focus on insights that will directly influence data cleaning, feature engineering, and model design.

**Examples to include:**
- Notable missing values or data quality issues.
- Feature distributions (skewed variables, outliers, etc.).
- Strong correlations or interactions.
- Early hypotheses about predictive relationships.

---

## 🚀 Next Steps

| Task | Description |
|------|--------------|
| **1. Data Cleaning** | Handle missing values, outliers, and inconsistent data types. |
| **2. Feature Engineering** | Create meaningful features based on insights (e.g., ratios, flags, time-based fields). |
| **3. Data Splitting** | Prepare train/test sets for model training. |
| **4. Modeling** | Proceed to baseline model setup and evaluation. |

---

**Notebook:** `01_eda.ipynb`
**Next Notebook:** `02_feature_engineering.ipynb`
**Author:** Brice
**Last Updated:** November 2, 2025

---
